first draft of section1 for data science pipelines (#4)
RedHatQuickCourses/rhods-pipelines: modules/chapter1/pages/section1.adoc

* *Documentation*: Creating documentation for the entire data science pipeline, including data sources, data processing steps, model details, and evaluation metrics. This documentation is essential for reproducibility and collaboration.

* *Collaboration and Communication*: Collaborating with domain experts, stakeholders, and other team members to define project objectives, gather domain knowledge, and align the data science work with the business goals.

These activities are typically iterative and may involve going back and forth between different stages to refine the analysis and models. Data science pipelines can vary depending on the specific project and organization, but these general steps provide a framework for approaching data science tasks.

== Data Science Pipelines in OpenShift AI

From the perspective of a data scientist or ML engineer user experience, OpenShift AI provides several user interface capabilities trying to cover as many as possible from these data science activities. It includes the following

* *Data Science Pipelines Main View* - End users can find the pipelines to whom they have access under the data science pipelines menu / main view

> image::../images/dsps-main.png[Data Science Pipelines Menu / Main View]

Reviewer: Antora is not rendering the images cause the paths are incorrect. Remove the `../images` portion in this image and the remaining ones.

<img width="707" alt="image" src="https://github.com/RedHatQuickCourses/rhods-pipelines/assets/675470/05658e69-ec7a-42db-a336-01c05c13c900">

```suggestion
image::dsps-main.png[Data Science Pipelines Menu / Main View]
```

> image::../images/dsps-main.png[Data Science Pipelines Menu / Main View]

Reviewer 2: ```suggestion
image::dsps-main.png[Data Science Pipelines Menu / Main View]
```

* *Data Science Pipelines Runs* - When looking for more details, end users explore the runs of their data science pipelines

> image::../images/dsps-runs.png[Data Science Pipelines Runs]

Reviewer 2: ```suggestion
image::dsps-runs.png[Data Science Pipelines Runs]
```

* *Data Science Pipelines view within a Data Science Project* - While within their data science project, end users can visualise the runs and other details of pipelines that ran within that project.

> image::../images/dsps-in-ds-project.png[Data Science Pipelines view within a Data Science Project]

Reviewer 2: ```suggestion
image::dsps-in-ds-project.png[Data Science Pipelines view within a Data Science Project]
```

Those data scientists and ML engineers with access to and knowledge of containers and OpenShift, can also look "under the covers" and check their pipeline runs direct in OpenShift:

* *OpenShift Pipeline executing a Data Science Pipeline* - From the OpenShift Console, under Pipelines->PipelinesRuns one can see the underlying Tekton pipeline that runs in OCP, once a data scientist started a Data Science Pipeline in OpenShift AI

> image::../images/dsp-run-in-ocp-pipelines.png[OpenShift Pipeline executing a Data Science Pipeline]

Reviewer 2: ```suggestion
image::dsp-run-in-ocp-pipelines.png[OpenShift Pipeline executing a Data Science Pipeline]
```

== Data Science Pipelines Technologies

At a high level, here are the technologies used to implement Data Science Pipelines in OpenShift AI:

* *Elyra Pipelines* - A *JupyterLab* extension, which provides a visual editor for creating data science pipelines based on Jupyter notebooks as well as Python or R scripts. User can drag and drop code and create visually such pipelines

> image::../images/elyra-pipeline-offline-scoring.png[An Offline Scoring pipeline in Elyra]

Reviewer 2: ```suggestion
image::elyra-pipeline-offline-scoring.png[An Offline Scoring pipeline in Elyra]
```

> image::../images/elyra-pipeline-model-training.png[A Model Training pipeline in Elyra]

Reviewer 2: ```suggestion
image::elyra-pipeline-model-training.png[A Model Training pipeline in Elyra]
```

A data scientist can assign resources visually, including CPUs and GPUs, to each individual step in the pipeline.

> image::../images/elyra-pipeline-step-config-with-gpu.png[Configuring GPUs and other resources for an Elyra pipeline step]

Reviewer 2: ```suggestion
image::elyra-pipeline-step-config-with-gpu.png[Configuring GPUs and other resources for an Elyra pipeline step]
```

* *Kubeflow Pipelines* - Specialized data science pipelines engine which can translate an Elyra visual pipeline execution (or Kubeflow SDK calls) into a *Tekton* pipeline running in OpenShift.

> image::../images/elyra-pipeline-running.png[Starting an Elyra pipeline]

Reviewer 2: ```suggestion
image::elyra-pipeline-running.png[Starting an Elyra pipeline]
```

> image::../images/elyra-pipeline-job-started.png[Details of a started Elyra pipeline]

Reviewer 2: ```suggestion
image::elyra-pipeline-job-started.png[Details of a started Elyra pipeline]
```

> * *OpenShift Pipelines (based on Tekton)* - It executes each step in the pipeline as an individual container / job and ensures that each containers gets allocated the resources (GPU, CPU, memory) configured/requested by data scientists in Elyra (or via Kubeflow SDK), as long as they are available in the OpenShift cluster.

Reviewer 2: ```suggestion
* *OpenShift Pipelines (based on Tekton)* - It executes each step in the pipeline as an individual container and ensures that each container gets allocated the resources (GPU, CPU, memory) configured/requested by data scientists in Elyra (or via Kubeflow SDK), as long as they are available in the OpenShift cluster.
```

Technically, Tekton is based on pods, not jobs.

> * *OpenShift GitOps with ArgoCD for AI/ML flows* - While not part of the OpenShift AI but rather as part of OpenShift, it is worth mentioning that once ML models are created, GitOps can be leveraged to push them to other OpenShift instances (including OpenShift Edge devices) for being hosted on model servers for inference and/or embedded in intelligent apps.

Reviewer 2: ```suggestion
* *OpenShift GitOps with ArgoCD for automated model deployments* - While not part of the OpenShift AI but rather as part of OpenShift, it is worth mentioning that once ML models are created, GitOps can be leveraged to push them to other OpenShift instances (including OpenShift Edge devices) for being hosted on model servers for inference and/or embedded in intelligent apps.
```
